node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
                     .event(move |maybe_block| {
                         let maybe_box_block = maybe_block.map(Box::new);
-                        Event::GetBlockForFinalitySignaturesResult(block_hash, maybe_box_block)
+                        Event::GetBlockForFinalitySignaturesResult(fs, maybe_box_block)
                     })
             }
-            Event::GetBlockForFinalitySignaturesResult(block_hash, None) => {
-                let signers = self
-                    .pending_finality_signatures
-                    .values()
-                    .filter_map(|sigs| sigs.get(&block_hash))
-                    .map(|fs| format!("{}", fs.public_key))
-                    .collect_vec();
-                if !signers.is_empty() {
-                    // We have signatures for an unknown block. Print log messages.
-                    warn!(
-                       %block_hash, signers = %signers.join(", "),
-                       "received signatures for a block that was not found in storage"
+            Event::GetBlockForFinalitySignaturesResult(fs, maybe_block) => {
+                if let Some(block) = &maybe_block {
+                    assert_eq!(
+                        block.hash(),
+                        &fs.block_hash,
+                        "block loaded from storage should have a matching block hash."
                     )
                 }
-                Effects::new()
+                // Check if validator is bonded in the era in which the block was created in.
+                effect_builder
+                    .is_bonded_validator(fs.era_id, fs.public_key)
+                    .map(|is_bonded| {
+                        if is_bonded {
+                            Ok((maybe_block, fs, is_bonded))
+                        } else {
+                            // If it's not bonded in that era, we will check if it's bonded in the
+                            // future era.
+                            Err((maybe_block, fs))
+                        }
+                    })
             }
-            Event::GetBlockForFinalitySignaturesResult(_block_hash, Some(block)) => {
-                let old_count = block.proofs().len();
-                let (block, mut effects) =
-                    self.add_pending_finality_signatures(*block, effect_builder);
-                if block.proofs().len() > old_count {
-                    let block = Box::new(block);
-                    effects.extend(effect_builder.put_block_to_storage(block).ignore());
+            .result(
+                |(maybe_block, fs, is_bonded)| Event::IsBonded(maybe_block, fs, is_bonded),
+                |(maybe_block, fs)| Event::IsBondedFutureEra(maybe_block, fs),
+            ),
+            Event::IsBondedFutureEra(maybe_block, fs) => {
+                // This should be safe to do. `self.latest_block` will be None only if ew haven't
+                // finalized *any* block yet. which will be just after the Genesis
```

Comment:
```suggestion
                // This should be safe to do. `self.latest_block` will be None only if we haven't
                // finalized *any* block yet, which will be just after the Genesis
```
(Or uppercase `Which`.)

Reply:
Done in [`c0bd13f` (#683)](https://github.com/CasperLabs/casper-node/pull/683/commits/c0bd13f70e41208382667535631dd62c9a381c5c)